### DISCUSSION TITLE
How to calculate average color

### DISCUSSION THREAD

**Discussion by joohei** [gfx-rs/wgpu#2687] [Q&A]
I have a program that recreates an image from other images using evolution. I want to calculate the average rgba-value of the area of a random object on the target image inside a fragment shader. Basically check the average color value underneath the object. I also don't want to render these objects to the screen. I need the average color value to be calculated inside the fragment shader and I also want to give the image that color. However, I don't know if that is even possible and how one would do that. Any help is appreciated!

Here is an example image of an object:
![image](https://user-images.githubusercontent.com/99501993/169697818-ce89b7b3-a9b5-4b81-a3e5-607d3bdae251.png)

Here is my current shader.wgsl file (basically copied from the wgpu tutorial):
![image](https://user-images.githubusercontent.com/99501993/169697852-82bdc16b-9788-4025-8a98-8293613d9d1c.png)

**Comment by joohei** [comment]
Basically I want to make that image look red when the position of the image on the target image has red color values on it. I want to keep the original shape of the image, but maybe give it a background color and the alpha channels will make it appear red.

**Reply by kvark** [reply]
Average color value is basically what the highest mip level (of size 1x1) would contain. You can follow the [mipmap](https://github.com/gfx-rs/wgpu/tree/master/wgpu/examples/mipmap) example for the details on how downsampling works.

**Reply by joohei** [reply]
That is an idea, but what about if the texture's resolution is not a power of two? To my understanding, calculating mipmaps is much more complex in that case.

**Reply by joohei** [reply]
> Average color value is basically what the highest mip level (of size 1x1) would contain. You can follow the [mipmap](https://github.com/gfx-rs/wgpu/tree/master/wgpu/examples/mipmap) example for the details on how downsampling works.

I'm quite a beginner to GPU stuff and find it quite difficult to follow that code since there aren't many explanations. I'm not sure how to implement mipmapping to my code. In addition to that I don't know how it is possible to create a 1x1 mipmap from a texture with resolution of (122x122) for example.
